File: funnyvote/app/src/main/java/com/heaton/funnyvote/ui/votedetail/VoteDetailContract.java
package com.heaton.funnyvote.ui.votedetail;

import com.heaton.funnyvote.BasePresenter;
import com.heaton.funnyvote.BaseView;
import com.heaton.funnyvote.database.Option;
import com.heaton.funnyvote.database.User;
import com.heaton.funnyvote.database.VoteData;

import java.util.List;

public interface VoteDetailContract {
    interface View extends BaseView<Presenter> {
        void showLoadingCircle();

        void hideLoadingCircle();

        void showSortOptionDialog(VoteData data);

        void showPollPasswordDialog();

        void hidePollPasswordDialog();

        boolean isPasswordDialogShowing();

        void shakePollPasswordDialog();

        void shakeAddNewOptionPasswordDialog();

        void showAddNewOptionPasswordDialog(final String newOptionText);

        void hideAddNewOptionPasswordDialog();

        void showResultOption(int optionType);

        void showUnPollOption(int optionType);

        void showExitCheckDialog();

        void showVoteInfoDialog(VoteData data);

        void showTitleDetailDialog(VoteData data);

        void showCaseView();

        void updateFavoriteView(boolean isFavorite);

        void setUpAdMob(User user);

        void setUpViews(VoteData voteData, int optionType);

        void setUpSubmit(int optionType);

        void setUpOptionAdapter(VoteData data, int optionType, List<Option> optionList);

        void showHintToast(int res);

        void showMultiChoiceToast(int max, int min);

        void showMultiChoiceAtLeast(int min);

        void showMultiChoiceOverMaxToast(int max);

        void refreshOptions();

        void updateChoiceOptions(List<Long> choiceList);

        void updateExpandOptions(List<String> expandList);

        void showShareDialog(VoteData data);

        void showAuthorDetail(VoteData data);

        void moveToTop();

        void updateSearchView(List<Option> searchList, boolean isSearchMode);

    }

    interface Presenter extends BasePresenter {
        void searchOption(String newText);

        void favoriteVote();

        void pollVote(String password);

        void resetOptionChoiceStatus(long optionId, String optionCode);

        void resetOptionExpandStatus(String optionCode);

        void addNewOptionStart();

        void addNewOptionContentRevise(long optionId, String inputText);

        void addNewOptionCompleted(String password, String newOptionText);

        void removeOption(long optionId);

        void showVoteInfo();

        void showTitleDetail();

        void IntentToShareDialog();

        void IntentToAuthorDetail();

        void changeOptionType();

        void sortOptions(int sortType);

        void CheckSortOptionType();

    }
}

File: funnyvote/app/src/main/java/com/heaton/funnyvote/ui/votedetail/VoteDetailContract.kt
package com.heaton.funnyvote.ui.votedetail

import com.heaton.funnyvote.BasePresenter
import com.heaton.funnyvote.BaseView
import com.heaton.funnyvote.database.Option
import com.heaton.funnyvote.database.User
import com.heaton.funnyvote.database.VoteData

interface VoteDetailContract {
    interface View : BaseView<Presenter> {

        val isPasswordDialogShowing: Boolean
        fun showLoadingCircle()

        fun hideLoadingCircle()

        fun showSortOptionDialog(data: VoteData)

        fun showPollPasswordDialog()

        fun hidePollPasswordDialog()

        fun shakePollPasswordDialog()

        fun shakeAddNewOptionPasswordDialog()

        fun showAddNewOptionPasswordDialog(newOptionText: String)

        fun hideAddNewOptionPasswordDialog()

        fun showResultOption(optionType: Int)

        fun showUnPollOption(optionType: Int)

        fun showExitCheckDialog()

        fun showVoteInfoDialog(data: VoteData)

        fun showTitleDetailDialog(data: VoteData)

        fun showCaseView()

        fun updateFavoriteView(isFavorite: Boolean)

        fun setUpAdMob(user: User)

        fun setUpViews(voteData: VoteData, optionType: Int)

        fun setUpSubmit(optionType: Int)

        fun setUpOptionAdapter(data: VoteData, optionType: Int, optionList: List<Option>)

        fun showHintToast(res: Int)

        fun showMultiChoiceToast(max: Int, min: Int)

        fun showMultiChoiceAtLeast(min: Int)

        fun showMultiChoiceOverMaxToast(max: Int)

        fun refreshOptions()

        fun updateChoiceOptions(choiceList: List<Long>)

        fun updateCurrentOptionsOrder(optionList: List<Option>)

        fun updateExpandOptions(expandList: List<String>)

        fun showShareDialog(data: VoteData)

        fun showAuthorDetail(data: VoteData)

        fun moveToTop()

        fun updateSearchView(searchList: List<Option>, isSearchMode: Boolean)

    }

    interface Presenter : BasePresenter {
        fun searchOption(newText: String)

        fun favoriteVote()

        fun pollVote(password: String)

        fun resetOptionChoiceStatus(optionId: Long, optionCode: String)

        fun resetOptionExpandStatus(optionCode: String)

        fun addNewOptionStart()

        fun addNewOptionContentRevise(optionId: Long, inputText: String)

        fun addNewOptionCompleted(password: String, newOptionText: String?)

        fun removeOption(optionId: Long)

        fun IntentToVoteInfo()

        fun IntentToTitleDetail()

        fun IntentToShareDialog()

        fun IntentToAuthorDetail()

        fun changeOptionType()

        fun sortOptions(sortType: Int)

        fun CheckSortOptionType()

    }
}
